pkg/controller/certificaterequests/venafi/venafi.go

```diff
@@ -104,17 +112,48 @@ func (v *Venafi) Sign(ctx context.Context, cr *cmapi.CertificateRequest, issuerO
 		}
 	}
 
-	certPem, err := client.Sign(cr.Spec.CSRPEM, duration, customFields)
+	pickupID, exists := cr.ObjectMeta.Annotations[VenafiPickupIDAnnotation]
 
-	// Check some known error types
-	if err != nil {
-		switch err.(type) {
+	// check if the pickup ID annotation is there, if not set it up.
+	if !exists || pickupID == "" {
+		pickupID, err = client.RequestCertificate(cr.Spec.CSRPEM, duration, customFields)
+		// Check some known error types
+		if err != nil {
+			switch err.(type) {
 
-		case venafiinternal.ErrCustomFieldsType:
-			v.reporter.Failed(cr, err, "CustomFieldsError", err.Error())
-			log.Error(err, err.Error())
+			case venaficlient.ErrCustomFieldsType:
+				v.reporter.Failed(cr, err, "CustomFieldsError", err.Error())
+				log.Error(err, err.Error())
 
-			return nil, nil
+				return nil, nil
+
+			default:
+				message := "Failed to request venafi certificate"
+
+				v.reporter.Failed(cr, err, "RequestError", message)
+				log.Error(err, message)
+
+				return nil, err
+			}
+		}
+
+		v.reporter.Pending(cr, err, "IssuancePending", "Venafi certificate is requested")
+
+		if cr.ObjectMeta.Annotations == nil {
+			cr.ObjectMeta.Annotations = map[string]string{}
+		}
+		cr.ObjectMeta.Annotations[VenafiPickupIDAnnotation] = pickupID
+		_, err = v.cmClient.CertmanagerV1alpha2().CertificateRequests(cr.GetNamespace()).Update(ctx, cr, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, err
+		}
+
+		return nil, nil
+	}
+
+	certPem, err := client.RetreiveCertificate(pickupID, cr.Spec.CSRPEM, duration, customFields)
+	if err != nil {
+		switch err.(type) {
```

Comment:
Is the returned certificate of one of these types or does each implementation *wrap* one of these types, in which case should we be using `errors.Is` instead?
And I'd have thought these errors would be found in the vcert `verrors` package....but wouldn't be surprised if now.

Reply:
They are not in verrors, since these are typed we can use `errors.As()` but that doesn't add much value to this switch case